src/digest/mod.rs

```diff
@@ -109,10 +109,8 @@ impl Context {
                 &mut self.pending[self.num_pending..self.algorithm.block_len],
                 &data[..to_copy]);
 
-            unsafe {
-                (self.algorithm.block_data_order)(self.state.as_mut_ptr(),
-                                                  self.pending.as_ptr(), 1);
-            }
+            (self.algorithm.block_data_order)(&mut self.state,
```

Comment:
The premise of this change is wrong. On ARM/AAarch64/x86/x86-64, block_data_order for SHA-{256,385,512} is implemented in assembly language and probably won't ever get replaced with Rust code. The C implementations might get replaced with Rust code, but they are mostly just placeholders for implementations that don't have assembly language implementations. SHA-1 is the exception because it is a deprecated algorithm and so the assembly language code was deleted.

Accordingly, the Rust implementations of these functions will be the exceptional case and so I don't think we should add overhead to the common case where the assembly language implementations are being used.

Note: On Windows, the assembly language implementations are not enabled yet, because I need to find a way to avoid depending on Perl on Windows, and I haven't gotten around to it yet.